library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java

```diff
@@ -0,0 +1,91 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.chat.api;
+
+import net.minecraft.util.Identifier;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.base.api.event.Event;
+import org.quiltmc.qsl.chat.api.types.AbstractChatMessage;
+
+import java.util.EnumSet;
+
+/**
+ * An interface for chat events. This copies much of its behavior from {@link Event}, but in its own interface.
+ *
+ * @param <C> the type of the callback for this event to accept in registration
+ * @param <R> the return type of invoking this event
+ */
+public interface ChatEvent<C, R> {
+	/**
+	 * Invokes the event with the provided message and returns the result or null. A null result is usually the result of no callbacks
+	 * being attached to this event, but may occur for any other reason depending on the implementation.
+	 *
+	 * @param message the message for the event to process
+	 * @return the result of invoking this event, or null
+	 */
+	@Nullable R invoke(@NotNull AbstractChatMessage<?> message);
+
+	/**
+	 * Invokes the event with the provided message, but replaces the result with {@code ifNull} if the result would have been null.
+	 *
+	 * @param message the message for the event to process
+	 * @param ifNull a value that should be returned instead of null
+	 * @return the result of invoking this event, or {@code ifNull} if the result would be null
+	 */
+	R invokeOrElse(@NotNull AbstractChatMessage<?> message, @NotNull R ifNull);
+
+	/**
+	 * Register a callback for the event.
+	 *
+	 * @param types an {@link EnumSet} of {@link QuiltMessageType} to determine what chat events to receive
+	 * @param callback the callback to register
+	 * @see #register(Identifier, EnumSet, Object)
+	 */
+	void register(@NotNull EnumSet<QuiltMessageType> types, @NotNull C callback);
+
+
+	/**
+	 * Registers a callback to a specific phase of the event.
+	 *
+	 * @param phaseIdentifier the phase identifier
+	 * @param callback the callback to register
+	 */
+	void register(@NotNull Identifier phaseIdentifier, @NotNull EnumSet<QuiltMessageType> types, @NotNull C callback);
+
+	/**
+	 * Request that callbacks registered for one phase be executed before callbacks registered for another phase.
+	 * <p>
+	 * Incompatible ordering constraints such as cycles will lead to inconsistent behavior:
+	 * some constraints will be respected and some will be ignored. If this happens, a warning will be logged.
+	 *
+	 * @param firstPhase  the identifier of the phase that should run before the other. It will be created if it didn't exist yet
+	 * @param secondPhase the identifier of the phase that should run after the other. It will be created if it didn't exist yet
+	 */
+	void addPhaseOrdering(@NotNull Identifier firstPhase, @NotNull Identifier secondPhase);
```

Comment:
I'm wondering if it's possible to have `@Contract("$1, $1 -> throw")` or something in those lines, if it's possible that would be a nice addition.

Reply:
Theoretically `@Contract(param1, param1 -> fail)`, but `param<N>` is specifically documented as "The method _returns_ its first (second, third, and so on) argument", so it might not work.
(see https://www.jetbrains.com/help/idea/contract-annotations.html#syntax)

Reply:
Can confirm that doesnt work, and I cant find a replacement

Reply:
bummer, it could've been super nice

damn you JetBrains!